File: app/app/lib/server/pipeline/steps/label-detection.ts
import type { PipelineStepDefinition } from "../types";
import { VideoIntelligenceServiceClient } from "@google-cloud/video-intelligence";
import { parseGoogleServiceAccount } from "@/app/lib/server/google-cloud";
import { getPipelineStateForAsset } from "../store";

function timeOffsetToSeconds(offset?: { seconds?: number | { low: number } | null; nanos?: number | null }) {
  const rawSeconds = offset?.seconds;
  const seconds = rawSeconds && typeof rawSeconds === "object" ? rawSeconds.low : Number(rawSeconds ?? 0);
  const nanos = Number(offset?.nanos ?? 0);
  return seconds + nanos / 1_000_000_000;
}

interface LabelSegment {
  start: number;
  end: number;
  confidence: number;
}

interface LabelEntity {
  entityId: string;
  description: string;
  languageCode: string;
}

interface DetectedLabel {
  entity: LabelEntity;
  categories: LabelEntity[];
  segments: LabelSegment[];
  confidence: number;
}

export const labelDetectionStep: PipelineStepDefinition = {
  id: "label-detection",
  label: "Detect labels",
  description: "Identifies objects, locations, activities, and more using the Video Intelligence API.",
  supportedTypes: ["video"],
  autoStart: true,
  run: async ({ asset }) => {
    const videoClient = new VideoIntelligenceServiceClient({
      credentials: parseGoogleServiceAccount(),
    });

    const pipeline = await getPipelineStateForAsset(asset.id);
    const uploadStep = pipeline.steps.find((step) => step.id === "cloud-upload");
    const gcsUri = uploadStep?.metadata?.gcsUri as string | undefined;
    if (!gcsUri) {
      throw new Error("Cloud upload step must complete before label detection");
    }

    const request = {
      inputUri: gcsUri,
      features: ["LABEL_DETECTION"],
      videoContext: {
        labelDetectionConfig: {
          labelDetectionMode: "SHOT_AND_FRAME_MODE",
          frameConfidenceThreshold: 0.5,
          videoConfidenceThreshold: 0.5,
        },
      },
    };

    const [operation] = await videoClient.annotateVideo(request as any);
    const [result] = await operation.promise();
    const annotations = result.annotationResults?.[0];

    // Process segment-level labels (whole video)
    const segmentLabels: DetectedLabel[] = (annotations?.segmentLabelAnnotations ?? []).map((label: any) => ({
      entity: {
        entityId: label.entity?.entityId ?? "",
        description: label.entity?.description ?? "",
        languageCode: label.entity?.languageCode ?? "en",
      },
      categories: (label.categoryEntities ?? []).map((cat: any) => ({
        entityId: cat.entityId ?? "",
        description: cat.description ?? "",
        languageCode: cat.languageCode ?? "en",
      })),
      segments: (label.segments ?? []).map((seg: any) => ({
        start: timeOffsetToSeconds(seg.segment?.startTimeOffset),
        end: timeOffsetToSeconds(seg.segment?.endTimeOffset),
        confidence: Number(seg.confidence ?? 0),
      })),
      confidence: Math.max(...(label.segments ?? []).map((seg: any) => Number(seg.confidence ?? 0)), 0),
    }));

    // Process shot-level labels
    const shotLabels: DetectedLabel[] = (annotations?.shotLabelAnnotations ?? []).map((label: any) => ({
      entity: {
        entityId: label.entity?.entityId ?? "",
        description: label.entity?.description ?? "",
        languageCode: label.entity?.languageCode ?? "en",
      },
      categories: (label.categoryEntities ?? []).map((cat: any) => ({
        entityId: cat.entityId ?? "",
        description: cat.description ?? "",
        languageCode: cat.languageCode ?? "en",
      })),
      segments: (label.segments ?? []).map((seg: any) => ({
        start: timeOffsetToSeconds(seg.segment?.startTimeOffset),
        end: timeOffsetToSeconds(seg.segment?.endTimeOffset),
        confidence: Number(seg.confidence ?? 0),
      })),
      confidence: Math.max(...(label.segments ?? []).map((seg: any) => Number(seg.confidence ?? 0)), 0),
    }));

    // Process frame-level labels (sampled at 1 fps)
    const frameLabelMap = new Map<string, { entity: LabelEntity; frames: Array<{ time: number; confidence: number }> }>();
    for (const label of annotations?.frameLabelAnnotations ?? []) {
      const entity: LabelEntity = {
        entityId: (label as any).entity?.entityId ?? "",
        description: (label as any).entity?.description ?? "",
        languageCode: (label as any).entity?.languageCode ?? "en",
      };
      const frames = ((label as any).frames ?? []).map((frame: any) => ({
        time: timeOffsetToSeconds(frame.timeOffset),
        confidence: Number(frame.confidence ?? 0),
      }));
      frameLabelMap.set(entity.description, { entity, frames });
    }
    const frameLabels = Array.from(frameLabelMap.values());

    // Sort by confidence
    segmentLabels.sort((a, b) => b.confidence - a.confidence);
    shotLabels.sort((a, b) => b.confidence - a.confidence);

    return {
      status: "succeeded" as const,
      metadata: {
        segmentLabelCount: segmentLabels.length,
        shotLabelCount: shotLabels.length,
        frameLabelCount: frameLabels.length,
        segmentLabels: segmentLabels.slice(0, 50), // Limit to top 50
        shotLabels: shotLabels.slice(0, 50),
        frameLabels: frameLabels.slice(0, 30),
        gcsUri,
      },
    };
  },
};

File: asset-service/src/asset_service/pipeline/steps/label_detection.py
"""Label detection pipeline step using Google Video Intelligence API."""

from __future__ import annotations

import logging
from typing import Any

from google.cloud import videointelligence_v1 as videointelligence
from google.oauth2 import service_account

from ..registry import register_step
from ..types import AssetType, PipelineContext, PipelineResult, StepStatus
from ..store import get_pipeline_state
from ...config import get_settings

logger = logging.getLogger(__name__)


def _time_offset_to_seconds(offset) -> float:
    """Convert protobuf duration to seconds."""
    if offset is None:
        return 0.0
    seconds = getattr(offset, "seconds", 0) or 0
    nanos = getattr(offset, "nanos", 0) or 0
    return float(seconds) + float(nanos) / 1_000_000_000


def _get_video_client():
    """Get Video Intelligence client with credentials."""
    settings = get_settings()
    key_path = settings.google_service_account_key or settings.firebase_service_account_key

    if key_path:
        from pathlib import Path
        import json

        path = Path(key_path).expanduser()
        if path.exists():
            credentials = service_account.Credentials.from_service_account_file(str(path))
        else:
            key_data = json.loads(key_path)
            credentials = service_account.Credentials.from_service_account_info(key_data)

        return videointelligence.VideoIntelligenceServiceClient(credentials=credentials)

    return videointelligence.VideoIntelligenceServiceClient()


def _parse_entity(entity) -> dict[str, Any]:
    """Parse a label entity."""
    return {
        "entityId": getattr(entity, "entity_id", "") or "",
        "description": getattr(entity, "description", "") or "",
        "languageCode": getattr(entity, "language_code", "en") or "en",
    }


def _parse_label_annotation(annotation) -> dict[str, Any]:
    """Parse a label annotation into our format."""
    entity = _parse_entity(annotation.entity) if annotation.entity else {}

    categories = [
        _parse_entity(cat) for cat in (annotation.category_entities or [])
    ]

    segments = []
    max_confidence = 0.0
    for seg in annotation.segments or []:
        segment_data = {
            "start": _time_offset_to_seconds(getattr(seg.segment, "start_time_offset", None)),
            "end": _time_offset_to_seconds(getattr(seg.segment, "end_time_offset", None)),
            "confidence": float(getattr(seg, "confidence", 0) or 0),
        }
        segments.append(segment_data)
        max_confidence = max(max_confidence, segment_data["confidence"])

    return {
        "entity": entity,
        "categories": categories,
        "segments": segments,
        "confidence": max_confidence,
    }


def _parse_frame_labels(frame_annotations) -> list[dict[str, Any]]:
    """Parse frame-level label annotations."""
    frame_label_map: dict[str, dict] = {}

    for annotation in frame_annotations or []:
        entity = _parse_entity(annotation.entity) if annotation.entity else {}
        description = entity.get("description", "")

        frames = []
        for frame in annotation.frames or []:
            frames.append({
                "time": _time_offset_to_seconds(frame.time_offset),
                "confidence": float(getattr(frame, "confidence", 0) or 0),
            })

        if description:
            frame_label_map[description] = {"entity": entity, "frames": frames}

    return list(frame_label_map.values())


@register_step(
    id="label-detection",
    label="Detect labels",
    description="Identifies objects, locations, activities, and more using the Video Intelligence API.",
    auto_start=True,
    supported_types=[AssetType.VIDEO],
)
async def label_detection_step(context: PipelineContext) -> PipelineResult:
    """Detect labels in video."""
    # Get GCS URI from upload step
    state = await get_pipeline_state(context.user_id, context.project_id, context.asset.id)
    upload_step = next((s for s in state.get("steps", []) if s["id"] == "cloud-upload"), None)
    gcs_uri = upload_step.get("metadata", {}).get("gcsUri") if upload_step else None

    if not gcs_uri:
        raise ValueError("Cloud upload step must complete before label detection")

    # Create client and request
    client = _get_video_client()

    label_config = videointelligence.LabelDetectionConfig(
        label_detection_mode=videointelligence.LabelDetectionMode.SHOT_AND_FRAME_MODE,
        frame_confidence_threshold=0.5,
        video_confidence_threshold=0.5,
    )

    video_context = videointelligence.VideoContext(
        label_detection_config=label_config,
    )

    request = videointelligence.AnnotateVideoRequest(
        input_uri=gcs_uri,
        features=[videointelligence.Feature.LABEL_DETECTION],
        video_context=video_context,
    )

    # Execute
    operation = client.annotate_video(request=request)
    result = operation.result(timeout=600)

    # Parse results
    segment_labels = []
    shot_labels = []
    frame_labels = []

    if result.annotation_results:
        annotations = result.annotation_results[0]

        # Segment-level labels (whole video)
        segment_labels = [
            _parse_label_annotation(ann)
            for ann in (annotations.segment_label_annotations or [])
        ]
        segment_labels.sort(key=lambda x: x["confidence"], reverse=True)

        # Shot-level labels
        shot_labels = [
            _parse_label_annotation(ann)
            for ann in (annotations.shot_label_annotations or [])
        ]
        shot_labels.sort(key=lambda x: x["confidence"], reverse=True)

        # Frame-level labels
        frame_labels = _parse_frame_labels(annotations.frame_label_annotations)

    return PipelineResult(
        status=StepStatus.SUCCEEDED,
        metadata={
            "segmentLabelCount": len(segment_labels),
            "shotLabelCount": len(shot_labels),
            "frameLabelCount": len(frame_labels),
            "segmentLabels": segment_labels[:50],  # Limit to top 50
            "shotLabels": shot_labels[:50],
            "frameLabels": frame_labels[:30],
            "gcsUri": gcs_uri,
        },
    )
